src/PathMatcher/CompositeMatcher.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Files\PathMatcher;
+
+use InvalidArgumentException;
+use function get_class;
+use function is_object;
+
+/**
+ * Composite matcher allows combining several matchers.
+ */
+final class CompositeMatcher implements PathMatcherInterface
+{
+    private bool $matchAny;
+
+    /**
+     * @var array PathMatcherInterface[]
+     */
+    private array $matchers;
+
+    private function __construct(bool $matchAny, array $matchers)
```

Comment:
```suggestion
    private function __construct(bool $matchAny, PathMatcherInterface ...$matchers)
```
I think typed variadic is more convenient way to check list items type